hazelcast/db.py

```diff
@@ -0,0 +1,509 @@
+import enum
+import itertools
+import threading
+import urllib.parse
+
+from hazelcast import HazelcastClient
+from hazelcast.config import Config
+from hazelcast.sql import (
+    HazelcastSqlError,
+    SqlColumnType,
+    SqlResult,
+    SqlRow,
+    SqlRowMetadata,
+    SqlExpectedResultType,
+    _DEFAULT_CURSOR_BUFFER_SIZE,
+)
+
+apilevel = "2.0"
+# Threads may share the module and connections.
+threadsafety = 2
+paramstyle = "qmark"
+
+DescriptionColumn = namedtuple(
+    "DescriptionColumn",
+    [
+        "name",
+        "type",
+        "display_size",
+        "internal_size",
+        "precision",
+        "scale",
+        "null_ok",
+    ],
+)
+
+
+class Type(enum.Enum):
+    NULL = 0
+    STRING = 1
+    BOOLEAN = 2
+    DATE = 3
+    TIME = 4
+    DATETIME = 5
+    INTEGER = 6
+    FLOAT = 7
+    DECIMAL = 8
+    JSON = 9
+    OBJECT = 10
+
+
+Date = date
+Time = time
+Timestamp = datetime
+Binary = bytes
+STRING = Type.STRING
+DATETIME = Type.DATETIME
+# The following identifiers exist just to satisfy the DBAPI 2 interface.
+# They are not actually used.
+BINARY = None
+NUMBER = None
+ROWID = None
+
+
+def DateFromTicks(ticks):
+    return date(*localtime(ticks)[:3])
+
+
+def TimeFromTicks(ticks):
+    return time(*localtime(ticks)[3:6])
+
+
+def TimestampFromTicks(ticks):
+    return datetime(*localtime(ticks)[:6])
+
+
+class RowResult:
+    def __init__(self, result):
+        self._result = result
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+
+class Cursor:
+    def __init__(self, conn: "Connection"):
+        self.arraysize = 1
+        self._conn = conn
+        self._res: Union[SqlResult, None] = None
+        self._description: Union[List[DescriptionColumn], None] = None
+        self._iter: Optional[Iterator[SqlRow]] = None
+        self._rownumber = -1
+        self._closed = False
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def __iter__(self) -> Optional[Iterator[SqlRow]]:
+        return self._iter
+
+    @property
+    def connection(self):
+        return self._conn
+
+    @property
+    def description(self) -> Union[List[DescriptionColumn], None]:
+        return self._description
+
+    @property
+    def rowcount(self) -> int:
+        return -1
+
+    @property
+    def rownumber(self) -> Optional[int]:
+        if self._rownumber < 0:
+            return None
+        return self._rownumber
+
+    def close(self):
+        if not self._closed:
+            self._closed = True
+            self._conn._close_cursor(self)
+            if self._res:
+                self._res.close()
+                self._res = None
+
+    def execute(self, operation: str, params: Optional[Tuple] = None) -> None:
+        if params is not None and not isinstance(params, tuple):
+            raise InterfaceError("params must be a tuple or None")
+        params = params or ()
```

Comment:
why do we need this? at this point, we know it is a non-null tuple

Reply:
`params` may be None.

Reply:
OK,  I missed the `params is not None` part